[WIP] Remove segment from list and add undo/redo for segments

philippotto opened the pull request:
… removing a segment

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- abc

### Issues:
- fixes #

------
(Please delete unneeded items, merge only when none are left open)
- [ ] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

daniel-wer reviewed (commented):
Code looks very good to me :+1: 

- [x] During testing I encountered some inconsistent behavior related to branching off from an earlier undo state. In that case the undo stack doesn't seem to be correctly pruned and one version of the old branch remains on the stack instead of the version that was branched from. See this gif, but let me know if you can't follow what's happening:
![segments_undo_redo_confusion](https://user-images.githubusercontent.com/1702075/228273757-86f6aadf-9488-433d-9dd2-618d6e7759e7.gif)
- [ ] When removing a segment from the list which had an ad-hoc or precomputed mesh loaded and then undoing that, the mesh check mark is restored but the mesh is not. Maybe the mesh loading can be retriggered in that case? Otherwise, the check mark shouldn't be checked.
- [x] [Low-Pri] Unrelated to this PR, but maybe it's quick to fix. The cross hair and brush icon next to the segment ID are in italics as well, looking slightly weird.

philippotto commented:
Thanks for your review & testing! I fixed the issues you found and also refactored the code a bit which is why I would recommend to review the changes commit-by-commit (one of them just extracts a large chunk of code into a new module). I hope the comments are self-explanatory, otherwise, let me know! Took me quite a bit of diving today to get into the save saga again :diving_mask:

daniel-wer commented on frontend/javascripts/oxalis/model/sagas/undo_saga.ts:
```diff
@@ -386,6 +390,7 @@ export function* manageUndoStates(): Saga<never> {
         if (!action) {
           throw new Error("Unexpected action");
         }
+        shouldClearRedoState = true;
```
Why does the redo state always need to be cleared, even for actions that are not added to the undo stack? :thinking: Should it have been moved to line 397 instead (so into the `if (addToUndo)` block)?

philippotto replied:
Yes, valid question! I added a comment for that. Let me know if you agree with the reasoning. The comment:

```
// Even if addToUndo is false, we will clear the redo state. Otherwise,
// selecting a segment while being in an earlier state and then doing a redo operation
// would forget about the newly registered segment which is weird.
// Clearing the redo state after an explicit action (such as selecting a segment)
// seems more intuitive.
// Also note that a segment update as a consequence of an annotation operation
// is already accompanied by an empty redo stack (because the volume operation
// will have cleared that stack). So, in that case shouldClearRedoState is true,
// but essentially nothing will be cleared, anyway.
```

daniel-wer replied:
Thank you for explaining. I agree with the reasoning, but am having some additional thoughts. I've actually wondered about this during the review, too, but shouldn't these other explicit segment update actions be undoable/redoable, then, too? It's a bit weird that something that clears the redo stack isn't undoable.
However, I also see the issue that the undo stack might get "crowded", but as long as only explicit user actions end up on the undo stack I would argue that's fine. Maybe in the future we can find a way to "highlight" where changes were undone or show a small description of what was just undone - which could help with that issue.

I don't think this needs to be fully discussed for this issue. The current state is fine to be merged.

philippotto replied:
> I agree with the reasoning, but am having some additional thoughts. I've actually wondered about this during the review, too, but shouldn't these other explicit segment update actions be undoable/redoable, then, too? It's a bit weird that something that clears the redo stack isn't undoable.

Ok, convinced :) Especially, the fact that changing a segment color wasn't undoable disturbed me. I just pushed a commit which takes care of this. I think it makes the code more explicit / easier, too. Could you do test the feature again? I did so, too, obviously, but I don't know whether I'm not seeing the forest for the trees after all this :)

daniel-wer reviewed (commented):
Please see my one question :) I'll test again once you say you're done with the PR.

daniel-wer commented on frontend/javascripts/components/color_picker.tsx:
```diff
@@ -47,19 +47,26 @@ export function ChangeColorMenuItemContent({
-    onSetColor(newColor);
+
+    // Only create a new undo state on the first color change event.
+    // All following color change events should mutate the most recent undo
+    // state so that the undo stack is not filled on each mouse movement.
```
This works because the color picker is unmounted when it is closed (and remounted when it's opened again), right?

philippotto replied:
Exactly! At first, I wanted to hook into the visibility-has-changed prop of the antd popover, but it turned out that wasn't even triggered, since the entire popover was simply unmounted when the menu closed. So, it's definitely unmounted :)

daniel-wer reviewed (approved):
Works wonderfully for me :tada:
